### DISCUSSION TITLE
Start and rollback transaction for the entirety of Jest suite

### DISCUSSION THREAD

**Discussion by tolgap** [mikro-orm/mikro-orm#2768] [Q&A] [answered]
One of the best things about Ruby On Rails (or to be more precise, ActiveRecord, the DB layer of Rails): wrapping your tests in transactions, and rolling them back after each test.

I was wondering if something like this could be achieved in MikroORM. Right now, I'm having to truncate my tables after each test and it just makes the entire test suite incredibly slow.

I was wondering how I could wrap `beforeEach` and `it` *in the same transaction*, and roll this transaction back in an `afterEach`. This should be applied to every test case, so it will most likely be in a `setupFilesAfterEnv`.

Question for @B4nan: If this is not possible, would it be worthwhile to allow MikroORM to instantiate a transaction, and manually set the transaction in your current EM. Something like this:

```ts
// transaction.setup.ts

import { Transaction } from '@mikro-orm/core';

let trx: Transaction;

beforeEach(async () => {
  // assume `global.orm` is available for the sake of this example
  trx = await global.orm.em.createTransaction();

  global.orm.em.setTransaction(trx);
});

afterEach(async () => {
  await trx.rollback();
  await global.orm.em.clearTransaction();
});
```

Looking forward to thoughts about an approach like this.

**Comment by B4nan** [comment, marked as answer]
You could try `em.begin()` and `em.rollback()`. You can also set the transaction context via `em.setTransactionContext()`, it is basically just a knex instance.

**Reply by tolgap** [reply]
Thanks! That looks like I can achieve manually beginning and rolling it back.

What happens if the code that I'm testing calls `persistAndFlush`. Would the transaction be committed? I'm trying to avoid committing any transactions.

Also it could be that it's not possible with MikroORM. I'm okay with "no" for an answer 👍 .

**Reply by B4nan** [reply]
Probably, never tried this approach myself. You can disable implicit transactions (those done when flushing) globally, via `implicitTransactions: false`. The transaction context will be still used:

https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/unit-of-work/UnitOfWork.ts#L331

**Reply by B4nan** [reply]
Btw if you can make this work, it would be a great article and a recipe in the docs :]

**Reply by tolgap** [reply]
@B4nan I managed to document my findings in this repository: https://github.com/tolgap/mikro-orm-transactional-tests

I managed to get it to work with `implicitTransactions: false`, `em.begin()`, `em.rollback()` and `em.clear()`. You can check it out if you want.

It works surprisingly well, even with the clumsiness of Jest `beforeEach` and `afterEach`.

**Comment by tkjwa** [comment]
Hello,

I'm working on similar use case in a slightly different context. I've noticed that in some cases, the data where not correctly rollbacked when using the approach describes by @tolgap 

1. direct usage of entity manager in tests for inserting datas leads to correct rollback after the test
2. usage of a seeder in tests for inserting datas leads to no rollback and duplicate errors if the same seeder is used in multiple tests
3. sending a request to the app (e2e test) leads to no rollback 

I've made a small repo to demonstrate the issue: https://github.com/jwa-lab/nest-mikro-orm/tree/not-working-as-expected _(edit: changed the link to the branch containing the version discussed here)_
run `npm i` then `docker-compose up -d` to start the db, then `npm run test:e2e` 

If I had to guess I'd say that's because using the Seeder and sending an http request will lead to a fork of the entity manager, which might somehow loose some transaction context 

My goal would be to have the following flow:
1. generate fresh db and run a "global" seeder from jest setup (not in the repo but works fine)
2. start a transaction before each test
3. run a "local" seeder inside the test if specific data are needed and then send the http request
4. rollback the transaction after each test (only the data from step 1 should remain in the db for next test)

Any help towards this will be highly appreciated !
Thank you very much

**Reply by B4nan** [reply]
`SeedManager` is always creating its own fork, and that will never share any TX context with anything else. Try using your seeders explicitly - instantiate them yourself and pass the EM fork you want - that way it will be used, only the `SeedManager` is forking, so you can skip that level completely.

Running a seeder is the same as instantiating it, calling `run(em)` on it and followed by `em.flush()` and `em.clear()` (which you might want to skip, depends on your use case).

edit: also `Seeder.call` is forking, you dont need to use that and again opt in for explicit seeder calls

**Reply by tkjwa** [reply]
Hello @B4nan , thank you for the feedback, I've been able to use the seeder as you suggested:
```
const seeder = new MySeeder();
await seeder.run(em);
```
Still no luck with the e2e test with supertest :(
I've been playing with transactions and setTransactionContext to a point where I get good query logs like
```
begin // begining of test suite
savepoint trx3 // begining of first test
// test queries
rollback to savepoint trx3 // end of first test
savepoint trx4 // begining of second test
// etc ...
```

Do you think it is somehow possible at all ?
Thank you

**Reply by tolgap** [reply]
@tkjwa you are doing `registerRequestContext: true` in your tests:

https://github.com/jwa-lab/nest-mikro-orm/blob/main/src/mikro-orm.config.ts#L12

It's hardcoded in your MikroORM config. You should set that to `false` when running your tests to avoid forking the em.

**Reply by B4nan** [reply]
Moreover that config is wrong, `registerRequestContext` is an option for the nestjs adapter, you cant just define it in your CLI config. This will have no effect and will be ignored.

> You should set that to false when running your tests to avoid forking the em.

Not sure what you mean by that, your tests (e2e ones) should use the context, as any other regular request.

> Do you think it is somehow possible at all ?

You fork before each test, as said above, this means each fork has its own (separate) transaction context. What you said so far seems like you don't want that.

https://github.com/jwa-lab/nest-mikro-orm/blob/main/test/app.e2e-spec.ts#L28

I also told you to _not_ use `SeedManager`, and you still do:

https://github.com/jwa-lab/nest-mikro-orm/blob/main/test/app.e2e-spec.ts#L69

**Reply by tolgap** [reply]
> Moreover that config is wrong, registerRequestContext is an option for the nestjs adapter, you cant just define it in your CLI config. This will have no effect and will be ignored.

I thought it was a `@mikro-orm/nestjs` feature that read the config, but apparently it's a core feature. I stand corrected.

> You fork before each test, as said above, this means each fork has its own (separate) transaction context. What you said so far seems like you don't want that.

This is what I meant to say. The only way I've been able to get this all to work in tests/e2e-tests is to only use the global EM, and start transactions in `beforeEach`, and rollback transactions + clearing the em in `afterEach`.

**Reply by tkjwa** [reply]
Hi guys, thank you very much for your help, it's really appreciated !

> Moreover that config is wrong, registerRequestContext is an option for the nestjs adapter, you cant just define it in your CLI config. This will have no effect and will be ignored.

I'll fix that, thanks for pointing it

> You fork before each test

If I don't, then I get the error about using the global em, which I can fix by setting allowGlobalContext: true

> I also told you to not use SeedManager, and you still do

Sorry I haven't updated the repo yet

> this means each fork has its own (separate) transaction context.

That make total sense now, sorry I missed that point ^^

> You should set that to false when running your tests to avoid forking the em.

Indeed, this is the config that is demonstrated in your repo

But then the tests run with a different config than the app when deployed, do you think that there are risks about missing errors due to using `allowGlobalContext: true` and `registerRequestContext: false` for the tests but not for running the app ?
I guess it would be really bad to run the app without `registerRequestContext: true`

Again, thanks for your help

**Comment by tkjwa** [comment]
Hi, I've managed to make everything work as I expected but it requires a lot of setup, some parts being weird, so I don't know if it's worth it.

The solution suggested by @tolgap is quite easy to setup but it requires to disable the `MikroOrmMiddleware` and to allow usage of global `EntityManager` during the tests.

The solution I came up to requires a lot more setup but allow to run the tests without any modification to the `MikroORM` config or the `AppModule`.

1. The transactions must be managed manually in the tests hooks, in `beforeAll` for [the test suite transaction](https://github.com/jwa-lab/nest-mikro-orm/blob/main/test/app.1.e2e-spec.ts#L40) and in `beforeEach` for [each test transaction](https://github.com/jwa-lab/nest-mikro-orm/blob/main/test/app.1.e2e-spec.ts#L56)
2. A new transaction is started for each test, and then passed to a [middleware](https://github.com/jwa-lab/nest-mikro-orm/blob/main/test/test.middleware.ts#L24) that have been added to the `Test.createTestingModule` call
3. When sending a request with supertest, the added middleware will run after the `MikroOrmMiddleware`, get the current `EntityManager` and feed it with the transaction from step 2
4. At the end of the test, the transaction can be rollbacked and any db change made through the app itself will be rollbacked

### The good parts
With this setup I am able to:

- use [jest global setup](https://jestjs.io/docs/configuration#globalsetup-string) to seed data available for the whole test suite (see [here](https://github.com/jwa-lab/nest-mikro-orm/blob/main/test/setup.ts))
- use a seeder in `beforeAll` hook for data that needs to be available for all tests in the file; the data are rollbacked at the end of the test suite (should also work in `beforeEach`) (see [here](https://github.com/jwa-lab/nest-mikro-orm/blob/main/test/app.1.e2e-spec.ts#L44))
- use a seeder in the test case for data specific to a single test; the data are rollbacked after the test (see [here](https://github.com/jwa-lab/nest-mikro-orm/blob/main/test/app.1.e2e-spec.ts#L84))
- send request with supertest to the Nest app; the data created this way are rollbacked after the test (see [here](https://github.com/jwa-lab/nest-mikro-orm/blob/main/test/app.1.e2e-spec.ts#L90))

### The bad parts
#### Nest middleware
It is not possible to fetch a Nest middleware instance from the container with something like `app.get(MyMiddleware);`
In order to "inject" the current transaction in the middleware I had to create a [service](https://github.com/jwa-lab/nest-mikro-orm/blob/main/test/test.middleware.ts#L5) (`CtxManager`) that can be injected in the middleware constructor and fetched with `app.get()` in the [jest hooks](https://github.com/jwa-lab/nest-mikro-orm/blob/main/test/app.1.e2e-spec.ts#L33)
#### Nest Module
As I didn't want to change the code of the `AppModule` because of the tests, I had to create a `TestModule` that is imported in the `beforeAll` hook when creating the app. This `TestModule` will be in charge to register the middleware.
```javascript
const module: TestingModule = await Test.createTestingModule({
  imports: [
    // Import the AppModule without any change to config
    AppModule,
    // Add the test module to register the TransactionContextMiddleware
    TestModule,
  ],
}).compile();
```
Unfortunately when registering the new middleware in this `TestModule`, it was running before the `MikroOrmMiddleware`, and couldn't fetch any `EntityManager` from the `RequestContext`.
In order to fix this I had to "bury" the middleware registration [deep](https://github.com/jwa-lab/nest-mikro-orm/blob/main/test/test.module.ts#L21) in a [module](https://github.com/jwa-lab/nest-mikro-orm/blob/main/test/test.module.ts#L16) [hierarchy](https://github.com/jwa-lab/nest-mikro-orm/blob/main/test/test.module.ts#L7), which seems really hacky ! see the [module file](https://github.com/jwa-lab/nest-mikro-orm/blob/main/test/test.module.ts)

### Conclusion

As I said in the begining, it's hard for me to decide if all of that is worth it, in comparison to the risks of running the tests without the `MikroOrmMiddleware` and with allowing the global `EntityManager`

**Reply by B4nan** [reply]
> Unfortunately when registering the new middleware in this TestModule, it was running before the MikroOrmMiddleware, and couldn't fetch any EntityManager from the RequestContext.
In order to fix this I had to "bury" the middleware registration [deep](https://github.com/jwa-lab/nest-mikro-orm/blob/main/test/test.module.ts#L21) in a [module](https://github.com/jwa-lab/nest-mikro-orm/blob/main/test/test.module.ts#L16) [hierarchy](https://github.com/jwa-lab/nest-mikro-orm/blob/main/test/test.module.ts#L7), which seems really hacky ! see the [module file](https://github.com/jwa-lab/nest-mikro-orm/blob/main/test/test.module.ts)

I believe this is a bug in nest.js, I also had troubles with unstable middleware execution, especially once you start nesting your modules.

**Reply by buzzware** [reply]
I am pleased to find this conversation and @tolgap  I still have a lot of love for Ruby on Rails even though I haven't used it for years.
@tkjwa I appreciate your valiant efforts to make it work without affecting the app ie like Rails and @B4nan I think this is an important goal. I've been trying to achieve this with Sqlite and had various problems - probably due to lack of nested transactions.
Jest could make this easier too - by enabling transactional test beforeAll/beforeEach hooks to be injected into a test suite.
Has any of this gotten easier or better since this thread ?

**Reply by buzzware** [reply]
With thanks to all I have got it working based on @tkjwa 's example here https://github.com/buzzware/nest-play
One key I found was to get the orm instance after module.compile() - it wasn't working when I created my own.
Also, it works without a transaction wrapping the whole suite - just test by test
